File: 프로그래머스/lv1_신규_아이디_추천.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

string solution(string new_id) {
    string answer = "";
    
    // 대문자로 변환 
    for(char c : new_id){
        c = tolower(c);
        // 특정문자만 선택 
        if( ('a' <= c && c <= 'z') || ('0' <= c && c <= '9') || c == '-' || c == '_' || c == '.' ){
            // 마침표가 2번 이상 연속이라면 삭제
            if ( !(answer.back() == '.' && c == '.') ){
                answer.push_back(c);
            }
        }
    }

    // 마침표가 처음이나 끝에 위치한다면 제거
    if ( answer.front() == '.' ){
        answer.erase(answer.begin());
    }
    if ( answer.back() == '.'){
        answer.erase(answer.end()-1);
    }
    // 빈문자열이라면 a 추가
    if ( answer.empty() ){
        answer.push_back('a');
    }

    // 길이가 16자 이상이면, 처음 15개의 문자를 제외한 나머지 문자열을 제거
    if ( answer.size() >= 16 ){
        answer.erase(answer.begin()+15, answer.end());
        // 제거한 문자열의 마지막이 . 이라면 제거
        if ( answer.back() == '.' ){
            answer.erase(answer.end()-1);
        }
    }
    
    // 문자열의 길이가 2보다 작다면, 길이가 3이 될때까지 뒤에 추가
    while (answer.size() <= 2){
        answer.push_back( answer.back() );
    }

    return answer;
}

int main(void){

    string new_id = "...!@BaT#*..y.abcdefghijklm";
    string ret;

    ret = solution(new_id);

    cout << ret << endl;
}

File: 프로그래머스/lv1_신규_아이디_추천.py

'''
    python 에서의 substring 은 [:] 슬라이싱으로 가능
    이때 뒤에오는 인자의 전 인덱스까지 수행함

    그리고 슬라이싱은 아웃오브인덱스를 검사하지 않음.
    범위를 벗어날 경우는 무시함

    str = "12345"
    str = str[-99:9]
    print(str) # 12345 출력

    풀이 후 답안을 확인하니, 정규표현식을 사용하면 쉽게 해결이 가능...
  
    import re

    def solution(new_id):
        st = new_id
        st = st.lower()
        st = re.sub('[^a-z0-9\-_.]', '', st)
        st = re.sub('\.+', '.', st)
        st = re.sub('^[.]|[.]$', '', st)
        st = 'a' if len(st) == 0 else st[:15]
        st = re.sub('^[.]|[.]$', '', st)
        st = st if len(st) > 2 else st + "".join([st[-1] for i in range(3-len(st))])
        return st
        
'''

def solution(new_id):
    answer = ''

    new_id = new_id.lower()

    for char in new_id:
        if( char.isdigit() or char.isalpha() or char == '-' or char == '_' or char == '.' ):
            continue
        else:
            new_id = new_id.replace(char, '')

    while( new_id.find("..") != -1 ):
        new_id = new_id.replace("..", ".")

    if( new_id[0] == '.'):
        new_id = new_id[1:]
    if( len(new_id) >= 1 and new_id[-1] == '.'):
        new_id = new_id[0:-1]

    if( len(new_id) == 0 ):
        new_id = 'a'
    
    if( len(new_id) >= 16 ):
        new_id = new_id[0:15]
        if( new_id[-1] == '.' ):
            new_id = new_id[0:-1]
    
    while( len(new_id) < 3 ):
        new_id += new_id[-1]

    answer = new_id
    return answer

if __name__ == "__main__":

    new_id = "abcdefghijklmn.p"
    ret = solution(new_id)
    print(ret)